### PR TITLE
feat: Firestore incremental capture extension

### PR DESCRIPTION
**Pull request by pr-Mais**
This PR adds a new extension built over PITR & DataFlow to back up and restore data in Firestore.

### CODE REVIEW

File: firestore-incremental-capture/extension.yaml [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    validationErrorMessage:
+      Must be a valid retention value, for example (7d or 10w)
+    default: 7d
+    required: true
+
+  - param: SYNC_DATASET
+    label: Bigquery dataset Id
+    description: >
+      The id of the Bigquery dataset to sync data to.
+
```
Comment (dackers86) [inline firestore-incremental-capture/extension.yaml:238]:
```suggestion
```

File: firestore-incremental-capture/functions/__tests__/backupDatabase.test.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import {runInitialSetup} from '../src/index';
+
+const mockQueue = jest.fn();
+
+const getFunctionsMock = () => ({
+  taskQueue: (functionName: string, instanceId: string) => ({
```
Comment (dackers86) [inline firestore-incremental-capture/functions/__tests__/backupDatabase.test.ts:6]:
```suggestion
  taskQueue: () => ({
```

File: firestore-incremental-capture/functions/__tests__/functions.test.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+process.env.FIREBASE_STORAGE_EMULATOR_HOST = '127.0.0.1:9199';
+
+/** Global vars */
+const {makeDocumentSnapshot} = fft.firestore;
+
+const db = admin.firestore();
+const collection = db.collection(config.syncCollectionPath);
+let randomId = '';
+
+xdescribe('functions', () => {
```
Comment (dackers86) [inline firestore-incremental-capture/functions/__tests__/functions.test.ts:37]:
```suggestion
describe('functions', () => {
```

File: firestore-incremental-capture/functions/__tests__/functions.test.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    /** generate random id */
+    randomId = (Math.random() + 1).toString(36).substring(7);
+
+    config.table = randomId;
+    config.dataset = randomId;
+
+    await initialize();
+  });
+
+  xit('Can sync data with BQ', async () => {
```
Comment (dackers86) [inline firestore-incremental-capture/functions/__tests__/functions.test.ts:52]:
```suggestion
  xit('Can sync data with BQ', async () => {
```

These tests may now be obsolete and can be removed.

File: firestore-incremental-capture/functions/__tests__/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import { DocumentReference, DocumentSnapshot } from "firebase-admin/firestore";
```
Comment (dackers86) [inline firestore-incremental-capture/functions/__tests__/helpers.ts:1]:
```suggestion
import { DocumentReference, DocumentSnapshot } from 'firebase-admin/firestore';
```

File: firestore-incremental-capture/functions/__tests__/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import { WrappedFirebaseFunction } from "./types";
```
Comment (dackers86) [inline firestore-incremental-capture/functions/__tests__/helpers.ts:2]:
```suggestion
import { WrappedFirebaseFunction } from './types';
```

File: firestore-incremental-capture/functions/src/dataflow/cloud_build.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    'mvn',
+    'compile',
+    'exec:java',
+    '-Dexec.mainClass=com.pipeline.RestorationPipeline',
+    `-Dexec.args=--runner=DataflowRunner --project=${config.projectId} --stagingLocation=${config.stagingLocation} --templateLocation=${config.templateLocation} --region=${config.dataflowRegion}`,
+  ],
+  dir: 'firebase-extensions/firestore-incremental-capture/functions/pipeline',
+};
+
+// const notifyStep = {
```
Comment (dackers86) [inline firestore-incremental-capture/functions/src/dataflow/cloud_build.ts:31]: Is this still needed?

File: firestore-incremental-capture/functions/src/dataflow/trigger_dataflow_job.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import * as admin from 'firebase-admin';
+import {logger} from 'firebase-functions/v1';
+import {FlexTemplatesServiceClient} from '@google-cloud/dataflow';
+// Cloud Firestore
```
Comment (dackers86) [inline firestore-incremental-capture/functions/src/dataflow/trigger_dataflow_job.ts:4]: Is this needed?

```suggestion
```

File: firestore-incremental-capture/functions/src/tasks/on_backup_restore_handler.ts [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  if (!isValidUnixTimestamp(timestamp)) {
+    logger.error(
+      '"timestamp" field is missing, please ensure that you are sending a valid timestamp in the request body, is in seconds since epoch and is not in the future.'
+    );
+    return Promise.resolve();
+  }
+
+  logger.info(`Running backup restoration at PIT ${timestamp}`);
+
+  // const importDoc = await db
```
Comment (dackers86) [inline firestore-incremental-capture/functions/src/tasks/on_backup_restore_handler.ts:17]: Is this code needed?

File: firestore-incremental-capture/functions/src/utils/big_query.ts [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+  /** Create a new table and return */
+  try {
+    logs.bigQueryTableCreating(tableId);
+
+    if (!dataset.id || !schema)
+      throw new Error('Dataset ID and schema must not be undefined');
+
+    /**
+     * TODO: Add time partitioning
```
Comment (dackers86) [inline firestore-incremental-capture/functions/src/utils/big_query.ts:60]: TODOs?

File: firestore-incremental-capture/functions/src/utils/big_query.ts [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+  return [dataset, table];
+}
+
+export async function getTable(datasetId: string, tableId: string) {
+  return bq.dataset(datasetId).table(tableId);
+}
+
+/**
+ * Export the Firestore db to storage
+ * TODO: This may now be obsolete. We can restore a database, and then replay the data through dataflow.
```
Comment (dackers86) [inline firestore-incremental-capture/functions/src/utils/big_query.ts:93]: TODO?

File: firestore-incremental-capture/POSTINSTALL.md [hunk 12]
```diff
@@ -0,0 +1,65 @@
+## Enable PITR in the Google Cloud Console
+
+## Creating a secondary Firestore database
```
[the remark this answers is not in the document]
Comment (pr-Mais) [reply firestore-incremental-capture/POSTINSTALL.md:3]: Thanks for noting, should fill in the docs here

**Review by dackers86 — APPROVED** [review]

**Pull request merged**